src/turnkeyml/common/filesystem.py

```diff
@@ -528,3 +554,198 @@ def rebase_cache_dir(input_path: str, build_name: str, new_cache_dir: str):
 
     relative_input_path = input_path.split(build_name, 1)[1][1:]
     return os.path.join(new_cache_dir, build_name, relative_input_path)
+
+
+def is_nice_to_write(value):
+    """
+    Checks whether a value is nice to write to YAML.
+    Returns True if the value is a string, int, float, bool, list, dict, or tuple.
+    Returns False otherwise.
+    """
+    if isinstance(value, (str, int, float, bool)):
+        return True
+    elif isinstance(value, list) or isinstance(value, tuple):
+        # Check if all elements in the list are nice to write
+        return all(is_nice_to_write(item) for item in value)
+    elif isinstance(value, dict):
+        # Check if all values in the dictionary are nice to write
+        return all(is_nice_to_write(item) for item in value.values())
+    return False
+
+
+def sanitize_for_yaml(input_dict: Dict) -> Dict:
+    """
+    Creates a new dictionary  containing only nice-to-write values
+    from the original dictionary.
+    """
+    result = {}
+    for key, value in input_dict.items():
+        if is_nice_to_write(value):
+            result[key] = value
+    return result
+
+
+class State:
+    """
+    The State class is meant to carry build state, starting with the user's
+    initial arguments, through each build Stage in the Sequence, and finally
+    to the disk, where it is used to assess cache hits.
+
+    State is initialized with the key members that are shared by every build,
+    and reasonable default values are assigned as appropriate.
+
+    Stage developers can also add any members they wish. To get or set an
+    attribute, reference it as an attribute:
+        1. get: `my_variable = state.attribute_name`
+        1. set: `state.attribute_name = my_variable`
```

Comment:
```suggestion
        2. set: `state.attribute_name = my_variable`
```
Python comments don't auto render